### PR TITLE
Rewrite the synthetic Go branch as master.

### PR DESCRIPTION
**Pull request by amscanne**
This enables legacy `go get` support, as well as vanilla `go get` with
modules. In theory, one should now be able to run:

```
go get gvisor.dev/runsc
go install gvisor.dev/runsc
```

And this should be always up-to-date.

### CODE REVIEW

File: cmd/gvisor-website/main.go [hunk 1]
```diff
@@ -41,20 +45,28 @@ var redirects = map[string]string{
 }
 
 var prefixHelpers = map[string]string{
-	"cl":     "https://gvisor-review.googlesource.com/c/gvisor/+/%s",
-	"change": "https://gvisor.googlesource.com/gvisor/+/%s",
```
Comment (iangudger) [inline cmd/gvisor-website/main.go]: Should we change these? If we do, we invalidate all of the existing links. Not the end of the world (there aren't too many), but I was thinking we could leave cl and change and add pr and pull for post-github migration changes. On the other hand, I can see how leaving the obsolete redirect could be confusing.

File: cmd/gvisor-website/main.go [hunk 2]
```diff
@@ -95,7 +107,7 @@ func hostRedirectHandler(h http.Handler) http.Handler {
 	})
 }
 
-// prefixRedirectHandler returns a handler that redirects to the given formated url.
```
Comment (iangudger) [inline cmd/gvisor-website/main.go]: Why remove the period?

**Comment by iangudger** [comment]
It looks like the go get paths in the commit message are out of date.

**Comment by prattmic** [comment]
This seems a bit complicated? We could just make the real master branch the generated code and move development to the "dev" branch. That makes things easy for users, and not much harder for developers. 

GitHub can even be configured to show the "dev" branch by default and have git clones default to that branch: https://help.github.com/en/articles/setting-the-default-branch I just double checked that it works for git clone.

**Comment by iangudger** [comment]
@prattmic, we could also just make a separate repository for the go getable version. We could even host it on googlesource.com given that we are going to be using the redirect.

**Comment by amscanne** [comment]
Yeah, I got nerd sniped on this and just wanted to make it work. Let's explore the ideas a little bit. I wonder how `go get` selects the master branch.

File: cmd/gvisor-website/main.go (hunk 1 above)
Comment (amscanne) [reply cmd/gvisor-website/main.go]: This is a good point. I think we should just add pr/ and pull/.

File: cmd/gvisor-website/main.go (hunk 2 above)
Comment (amscanne) [reply cmd/gvisor-website/main.go]: Accident. Will restore.

**Comment by amscanne** [comment]
> GitHub can even be configured to show the "dev" branch by default and have git clones default to that branch: https://help.github.com/en/articles/setting-the-default-branch I just double checked that it works for git clone.

Unfortunately the default for git clone is the problem here. That would also make the dev branch be the one that `go get` refers to by default. (See the clone call in [this file](https://golang.org/src/cmd/go/internal/get/vcs.go).)

I'd rather not maintain two different repositories, if we can help it. Let's think on it for a while.

File: cmd/gvisor-website/go.mod [hunk 3]
```diff
@@ -3,6 +3,55 @@ module main
 go 1.12
 
 require (
-	golang.org/x/oauth2 v0.0.0-20190226205417-e64efc72b421
+	github.com/BurntSushi/xgb v0.0.0-20160522181843-27f122750802 // indirect
```
[the remark this answers is not in the document]
Comment (amscanne) [reply cmd/gvisor-website/go.mod]: They are the calculated indirect dependencies. It should be checked in.

**Comment by amscanne** [comment]
I've updated the code here and made it much cleaner and more maintainable.

Effectively this target will simply change the *default* branch to the `go` branch. I think this is actually the best way to support `go get` while minimizing the number of repositories that we need to maintain. I don't rewrite any references here or drop anything. You can still develop from the checked out repository, assuming you checkout master.

File: cmd/gvisor-website/go.mod (hunk 3 above)
[the remark this answers is not in the document]
Comment (amscanne) [reply cmd/gvisor-website/go.mod]: Can you resolve this? Checking in the calculated go.mod and go.sum files is the preferred methodology AFAIK.

File: cmd/gvisor-website/main.go (hunk 1 above)
Comment (amscanne) [reply cmd/gvisor-website/main.go]: Note that I've added these but the comment does not resolve but the specific lines here did not change.

File: cmd/gvisor-website/main.go [hunk 4]
```diff
@@ -194,6 +215,231 @@ func registerRebuild(mux *http.ServeMux) {
...
+	proxy.ServeHTTP(w, r)
+}
+
+func infoRefs(w http.ResponseWriter, r *http.Request) {
+	// We intercept the client request. We implement only the appropriate
+	// git-upload-pack service and enforce that clients are asking for this
+	// as well. The full specification of this service, as well as a formal
+	// grammar is available at:
+	//
+	//	https://gist.github.com/schacon/6092633#file-git-http-proto-txt-L224
```
Comment (iangudger) [inline cmd/gvisor-website/main.go]: https://github.com/git/git/blob/master/Documentation/technical/http-protocol.txt seems to be the canonical version.

File: cmd/gvisor-website/main.go [hunk 5]
```diff
@@ -194,6 +215,231 @@ func registerRebuild(mux *http.ServeMux) {
...
+			fmt.Fprintf(w, "%04x", 0)
+		} else {
+			fmt.Fprintf(w, "%04x%s\n", 4+len(m)+1, m)
+		}
+	}
+
+	// readPkt reads a single packet line.
+	readPkt := func() (string, bool) {
+		// Parse the size header and return the string. Note that we
+		// don't do bounds checking here and rely on the recovery
```
Comment (iangudger) [inline cmd/gvisor-website/main.go]: Is this true? It looks like there is bounds checking going on and I don't see a recover.

File: cmd/gvisor-website/main.go [hunk 6]
```diff
@@ -194,6 +215,231 @@ func registerRebuild(mux *http.ServeMux) {
...
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintf(w, "invalid reference: %v", line)
+			return "", "", nil, false
+		}
+		parts := strings.Split(line, "\x00")
+		if len(parts) > 2 {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintf(w, "invalid reference: %v", line)
+			return
+		} else if len(parts) == 2 {
```
Comment (iangudger) [inline cmd/gvisor-website/main.go]: else isn't necessary

File: cmd/gvisor-website/main.go [hunk 7]
```diff
@@ -194,6 +215,231 @@ func registerRebuild(mux *http.ServeMux) {
...
+		if !ok || line == "" {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintf(w, "invalid reference: %v", line)
+			return "", "", nil, false
+		}
+		parts := strings.Split(line, "\x00")
+		if len(parts) > 2 {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintf(w, "invalid reference: %v", line)
+			return
```
Comment (iangudger) [inline cmd/gvisor-website/main.go]: It seems a little odd to switch between manually and implicitly specifying the return values. This case and the following don't seem materially different than the previous two.

File: cmd/gvisor-website/main.go (hunk 6 above)
Comment (ianlewis) [reply cmd/gvisor-website/main.go]: If len(parts) is less than 2 then the following line will panic, no? (The len(parts) == 1 case isn't handled afaict). Unless of course readPkt guarantees that a string of len > 0 and without a delimiter is never returned somehow.

File: cmd/gvisor-website/main.go (hunk 7 above)
Comment (ianlewis) [reply cmd/gvisor-website/main.go]: +1. implicit return values (esp. in a large function like this) are the devil

File: cmd/gvisor-website/main.go (hunk 6 above)
Comment (amscanne) [reply cmd/gvisor-website/main.go]: The len(parts) == 1 case is handled below. Maybe you meant len(parts) == 0? That's guaranteed not to occur per the strings.Split semantics and the fact that I check for an empty string.

File: cmd/gvisor-website/main.go (hunk 7 above)
Comment (amscanne) [reply cmd/gvisor-website/main.go]: Fixed.

File: cmd/gvisor-website/main.go [hunk 8]
```diff
@@ -30,31 +35,47 @@ import (
...
+	"/cl": "https://gvisor-review.googlesource.com",
 }
 
 var prefixHelpers = map[string]string{
-	"cl":     "https://gvisor-review.googlesource.com/c/gvisor/+/%s",
-	"change": "https://gvisor.googlesource.com/gvisor/+/%s",
+	"change": "https://github.com/google/gvisor/commit/%s",
 	"issue":  "https://github.com/google/gvisor/issues/%s",
+	"pr":     "https://github.com/google/gvisor/pull/%s",
+	"pull":   "https://github.com/google/gvisor/pull/%s",
```
[the remark this answers is not in the document]
Comment (amscanne) [reply cmd/gvisor-website/main.go]: I've dropped pull and just kept pr.

File: cmd/gvisor-website/main.go [hunk 9]
```diff
@@ -194,6 +215,231 @@ func registerRebuild(mux *http.ServeMux) {
...
+	}
+
+	mux.Handle("/gvisor/git-upload-pack", hostRedirectHandler(http.HandlerFunc(gitUploadPack)))
+	mux.Handle("/gvisor/info/refs", hostRedirectHandler(http.HandlerFunc(infoRefs)))
+}
+
+func gitUploadPack(w http.ResponseWriter, r *http.Request) {
+	target, err := url.Parse(upstreamGitUploadPack)
+	if err != nil {
+		panic(err) // Should never happen.
```
[the remark this answers is not in the document]
Comment (amscanne) [reply cmd/gvisor-website/main.go]: I've made this a static initializer.

File: cmd/gvisor-website/main.go (hunk 5 above)
Comment (amscanne) [reply cmd/gvisor-website/main.go]: This was a previous iteration and the comment was stale. I decided to make it cleaner without that. Remove obsolete comment.

File: cmd/gvisor-website/main.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ianlewis) [reply cmd/gvisor-website/main.go]: It looks like you dropped pr and kept pull? I was a bit partial to pr but I guess I could go either way

File: cmd/gvisor-website/main.go (hunk 6 above)
Comment (ianlewis) [reply cmd/gvisor-website/main.go]: Ah, ok, I misunderstood. The if is necessary but not the else.

**Comment by iangudger** [comment]
The title is still "Rewrite the synthetic Go branch as master." Is that still accurate?

**Comment by amscanne** [comment]
I've decided to delay this pull request in favor of #78. I'll leave it here for a while, but since we need go modules anyways, I don't think there's a huge amount of value in this indirection.

**Comment by iangudger** [comment]
Don't we also need a simpler gvisor.dev/gvisor redirect?

**Comment by amscanne** [comment]
So here's the summary:

The vanity package will never work with vanilla `go get`. It's because that will pull all the package dependencies from tip, and there are some fundamental things broken with that.

Since that's the case, go modules are essentially _required_. So I think I'll just leave things the way they are. You'll need to specify the `go` branch to use `go get`, but that should make it clear that's special and not a generally supported route.

**Pull request closed without merging**